CONTRIBUTING.md

```diff
@@ -0,0 +1,900 @@
+	while (hasNext()) {
+		list.push(next(index))
+		index++
+	}
+
+	for (var i = 0; i < list.length; i++) {
+		read(list[i])
+	}
+}
+
+// Bad
+function iterate() {
+	var index = 0
+	var list = []
+	while (hasNext()) list.push(next(index++))
+	for (var i = 0; i < list.length; i++) read(list[i])
+}
+
+// Even worse
+function iterate() {
+	var list = []
+	for (var index = 0; hasNext(); list.push(next(index++))) {}
+	for (var i = 0; i < list.length; i++) read(list[i])
+}
+```
+
+### Trailing whitespace
+
+Please don't leave trailing whitespace anywhere. It makes diffs harder to read.
+
+Blank lines should have no indentation or spaces. It should be empty other than a line ending.
+
+All files should end with a trailing line break. Some editors add this on their own, so it will result in unnecessary lines in later diffs if you don't.
+
+### Indentation and vertical whitespace
+
+Indent with hard tabs. Each one counts for 4 spaces.
+
+Never mix tabs and spaces. Don't use smart tabs.
+
+### Excessive whitespace
+
+Keep whitespace within reason. Limit vertical whitespace to no more than 2 consecutive blank lines, and don't start or end blocks with plain whitespace. Don't use more than one character of horizontal whitespace beyond indentation.
+
+```js
+// Bad
+function iterate() {
+	var index = 0
+	var list = []
+
+	while (hasNext()) {
+		list.push(next(index))
+		index++
+	}
+
+	// Too many empty lines
+
+
+
+	for (var i = 0; i < list.length; i++) {
+		read(list[i])
+	}
+}
+
+// Also bad
+var a  =  2
+//   ^^ ^^
+```
+
+### Control keywords
+
+Always surround control keywords (e.g. `if`, `else`, `for`) with whitespace.
+
+## Operators
+
+### Binary operators
+
+Always surround binary keyword operators (e.g. `in`, `instanceof`) with whitespace.
+
+Always surround any other binary operator with whitespace, including assignment operators. Add line breaks after the operator, not before.
+
+```js
+// Good
+var a = 1 + 2
+a = 3
+a += 4
+
+var a = 1 + 2 + 3 + 4 + 5 +
+	6 + 7 + 8 + 9 + 10
+
+// Bad
+var a = 1+2
+var a = 1 +2
+var a = 1+ 2
+var a=1+2
+var a=1 + 2
+a=3
+a+=4
+
+var a = 1 + 2 + 3 + 4 + 5
+	+ 6 + 7 + 8 + 9 + 10
+```
+
+As an exception, casting to a 32-bit integer (i.e. `x|0`) doesn't require whitespace surrounding it.
+
+```js
+// This is okay
+var casted = value|0
+```
+
+### Unary operators
+
+Always use whitespace between an unary keyword operator (e.g. `delete`, `new`) with whitespace.
+
+Do not use spaces between any other unary operator and the value they apply to.
+
+```js
+// Good
+!a
+!!a
+~~a
+a++
+++a
+-1
+~1
+
+// Okay
++!a
+
+// Bad
+! a
+!! a
+! ! a
+++ a
+a ++
++ ! a
+
+// Even worse
+- 1
+~ 1
+```
+
+Don't use `new function () {}`. Just use an object literal instead, and an IIFE if needed.
+
+## Objects
+
+### Exterior whitespace
+
+Do not include space between the opening curly brace and the first object key in single line objects. For multi-line objects, don't include any extra preceding whitespace at the beginning or end.
```

Comment:
For what it's worth, this is the first style guide I've written. Sorry if
it's not quite perfect. (typed from phone)

On Tue, Nov 10, 2015, 17:51 Richard Eames <notifications@github.com> wrote:

> In CONTRIBUTING.md
> <https://github.com/lhorie/mithril.js/pull/834#discussion_r44478142>:
>
> > +++ a
> > +a ++
> > ++ ! a
> > +
> > +// Even worse
> > +- 1
> > +~ 1
> > +```
> > +
> > +Don't use `new function () {}`. Just use an object literal instead, and an IIFE if needed.
> > +
> > +## Objects
> > +
> > +### Exterior whitespace
> > +
> > +Do not include space between the opening curly brace and the first object key in single line objects. For multi-line objects, don't include any extra preceding whitespace at the beginning or end.
>
> Also, not to include whitespace between the value, and the closing curly
> brace.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lhorie/mithril.js/pull/834/files#r44478142>.
>